### PR TITLE
Generalize file-based RPC command client beyond VSCode

### PR DESCRIPTION
**Pull request by JohnEffo**
# What is this

Split command-client so that other application can reuse the logic for of command server for RPC using existing VS code Command-Client.

## Reasoning
In slack advised by aegis, when discussing RPC mechanism for Visual Studio.

"
if you replicate the vscode file rpc, please try to port the existing code as precisely as possible, without changing the sequence of file reads/writes or anything, because we spent a while making that work well
"

Initially copied and pasted existing code, wrote the Visual Studio side of things, then thought that best port is re-use of existing code

## Discussion
* If this PR goes ahead is Command-Client in the correct location? Probably not, but was not sure were to place it and wanted to float this idea before continuing with anymore work.
* Should the methods not be renamed so that they are not so VSCode specific? Yes probably, for the moment I'm happy with the way things work and there is existing code which would need updating (Cursorless) which I did not want to touch as part of this PR. 

## Apologies

I've installed pre-commit but have not gotten it working correctly as of yet, but wanted to get your opinion on, whether this is this a valid direction of travel.

### CODE REVIEW

File: apps/vscode/command_client/command_client.py
Comment (pokey) [inline apps/vscode/command_client/command_client.py:365]:
```suggestion
        """The strategy to use when no command server directory is located for the the application"""
```

File: apps/vscode/command_client/command_client.py
Comment (pokey) [inline apps/vscode/command_client/command_client.py:260]: How did these get deleted?

File: apps/vscode/command_client/command_client.py
Comment (pokey) [inline apps/vscode/command_client/command_client.py:230]: this one too

File: apps/vscode/command_client/command_client.py
Comment (pokey) [inline apps/vscode/command_client/command_client.py:213]: and this

File: apps/vscode/command_client/command_client.py
Comment (pokey) [inline apps/vscode/command_client/command_client.py:142]: more

File: apps/vscode/command_client/command_client.py
Comment (pokey) [inline apps/vscode/command_client/command_client.py:95]: again

File: apps/vscode/command_client/command_client.py
Comment (pokey) [inline apps/vscode/command_client/command_client.py:61]: one more 😅

File: apps/vscode/command_client/vs_code.py
Comment (pokey) [inline apps/vscode/command_client/vs_code.py:29]: What is this supposed to do?  This is causing the following error when I try to run pretty much any command in VSCode (other than Cursorless commands, which don't use this function) 😅:

```
KeyError: "Action 'user.post_command' is not declared."
```

**Review by pokey — CHANGES_REQUESTED** [review]

**Comment by pokey** [comment]
Good stuff!  Left a bunch of comments but I think overall this is a nice direction

In the future we should probably have an action called `run_command` instead of expecting all applications to just use an action called `vscode` 😅.  Could probably add that later 🤷

File: apps/vscode/command_client/visual_studio.py
Comment (JohnEfford) [reply apps/vscode/command_client/visual_studio.py:9]: Yes, originally I had this in command_client and never renamed when I copied in out.

File: apps/vscode/command_client/vs_code.py
Comment (JohnEfford) [reply apps/vscode/command_client/vs_code.py:9]: Done

File: apps/vscode/command_client/command_client.py
Comment (JohnEfford) [reply apps/vscode/command_client/command_client.py:38]: I think this is what I was striving for for but could no make the final leap.

File: apps/vscode/command_client/command_client.py
Comment (JohnEfford) [reply apps/vscode/command_client/command_client.py:345]: Done

File: apps/vscode/command_client/command_client.py
Comment (JohnEfford) [reply apps/vscode/command_client/command_client.py:356]: Done

File: apps/vscode/command_client/command_client.py
Comment (JohnEfford) [reply apps/vscode/command_client/command_client.py:365]: Done

File: apps/vscode/command_client/vs_code.py
Comment (JohnEfford) [reply apps/vscode/command_client/vs_code.py:29]: Removed I think it was part of embroyonic attempt to have both new command_server command as vscommands working for backwards compatibility, before I decided that that could wait.

File: apps/vscode/command_client/command_client.py
Comment (JohnEfford) [reply apps/vscode/command_client/command_client.py:260]: Hmm not sure, I've added them back.

File: apps/vscode/command_client/command_client.py
Comment (JohnEfford) [reply apps/vscode/command_client/command_client.py:230]: Done

File: apps/vscode/command_client/command_client.py
Comment (JohnEfford) [reply apps/vscode/command_client/command_client.py:213]: Done

File: apps/vscode/command_client/command_client.py
Comment (JohnEfford) [reply apps/vscode/command_client/command_client.py:95]: Done

File: apps/vscode/command_client/command_client.py
Comment (JohnEfford) [reply apps/vscode/command_client/command_client.py:61]: Done

File: apps/vscode/command_client/command_client.py
Comment (pokey) [inline apps/vscode/command_client/command_client.py:147]: The idea would be to throw an exception here

File: apps/vscode/command_client/vs_code.py
Comment (pokey) [inline apps/vscode/command_client/vs_code.py:17]: And then define the action `vscode` here, and call the `run_command` action, wrapped in a try-catch. If it throws an error, do the fallback

File: apps/vscode/command_client/vs_code.py
Comment (pokey) [inline apps/vscode/command_client/vs_code.py:18]: Then you can remove this action; just inline into `vscode` above

File: apps/vscode/command_client/vs_code.py
Comment (pokey) [inline apps/vscode/command_client/vs_code.py:1]: This file should be called `vscode.py`

**Review by pokey — COMMENTED** [review]
Getting close. As discussed, it would be good to

- [ ] rename the actions starting with `vscode_` to `run_command_*`
- [ ] Then add shims in vscode that call these actions
- [ ] Open a PR in cursorless that tries the `run_command_` versions, falling back to `vscode_` if action not found

Also see below for the idea I had around raising an exception

File: apps/vscode/command_client/command_client.py
Comment (JohnEfford) [reply apps/vscode/command_client/command_client.py]: Done

File: apps/vscode/command_client/vs_code.py
Comment (JohnEfford) [reply apps/vscode/command_client/vs_code.py]: Done, this but duplicated code

File: apps/vscode/command_client/vs_code.py
Comment (JohnEfford) [reply apps/vscode/command_client/vs_code.py]: Done

File: apps/vscode/command_client/command_client.py
Comment (rntz) [reply apps/vscode/command_client/command_client.py:95]: doesn't look fixed? still deleting lines here

File: apps/vscode/command_client/command_client.py
Comment (rntz) [inline apps/vscode/command_client/command_client.py:122]: while `run_vscode_command` is now too specific, `user.run_command` is a bit too generic. Maybe `editor_command`? `application_command`?

File: apps/vscode/command_client/command_client_tag.py
Comment (rntz) [inline apps/vscode/command_client/command_client_tag.py:23]: I don't really understand why this is a separate file from `command_client.py`, or how to determine what goes in which file.

**Review by rntz — COMMENTED** [review]
I pushed a few changes to this PR, I hope you don't mind. In particular:

- I removed some changes that you made to the formatters that I assume weren't intended to be part of this PR.

- I restored some empty lines you deleted.

- I fixed the type annotation on `*args` in `run_command` and added this to its docstring; and moved the default implementation of `emit_pre_phrase_signal` as `return False` from the global context to the module where it's defined. This could have been a separate PR but they would conflict, so I thought it would be easier for pokey to review them together.

**Comment by rntz** [comment]
Looks good! I hope this lands soon, I will probably need this to get cursorless-in-emacs working.

File: apps/vscode/command_client/command_client.py
Comment (JohnEfford) [reply apps/vscode/command_client/command_client.py:151]: Done

File: apps/vscode/command_client/command_client_tag.py
Comment (JohnEfford) [reply apps/vscode/command_client/command_client_tag.py:23]: I thought that all tags should reside in there own file. I put the Actions here because these are the actions a user command_client would need to implement as shown in the visual_studio.py file. I thought putting them in command_client would make it more difficult to assertain what needed to be implemeted for another system. I'm definitly a Newbie when it comes to Python and Talon scripting, so any guidence whould be appreciated.

File: apps/vscode/command_client/command_client.py
Comment (JohnEfford) [reply apps/vscode/command_client/command_client.py:287]: run_command was suggested but, having tried to use it in a Cursorless environment I agree it is two generic. run_rpc_command does exaclty what it says on the tin.

Comment (JohnEffo) [reply apps/vscode/command_client/command_client.py]: @pokey I removed the single parameter version's of this command there only available to vscode via vscode.py.

File: apps/vscode/command_client/command_client.py
Comment (rntz) [inline apps/vscode/command_client/command_client.py:3]: I think this shouldn't be here. You seem to be using this as a type, but it isn't a type; the Python type for strings is `str`.

File: apps/vscode/command_client/command_client.py
Comment (rntz) [inline apps/vscode/command_client/command_client.py:343]: should be `str` not `string`

File: apps/vscode/command_client/command_client_tag.py
Comment (rntz) [inline apps/vscode/command_client/command_client_tag.py:1]: Should be removed.

File: apps/vscode/command_client/visual_studio.py
Comment (rntz) [inline apps/vscode/command_client/visual_studio.py:1]: Should be removed.

File: apps/vscode/command_client/vscode.py
Comment (rntz) [inline apps/vscode/command_client/vscode.py:1]: Should be removed.

File: apps/vscode/command_client/vscode.py
Comment (rntz) [inline apps/vscode/command_client/vscode.py:33]: Should be `str`.

**Review by rntz — CHANGES_REQUESTED** [review]

File: apps/vscode/command_client/command_client.py
Comment (JohnEfford) [reply apps/vscode/command_client/command_client.py:343]: Done

File: apps/vscode/command_client/command_client.py
Comment (JohnEfford) [reply apps/vscode/command_client/command_client.py:3]: Thank you for the heads up.

**Comment by pokey** [comment]
Have you maybe forgotten to push your latest changes?  There seem to be a number of places that you've said "Done", but it doesn't look done 😅 (eg https://github.com/knausj85/knausj_talon/pull/956/files#r954160395 and https://github.com/knausj85/knausj_talon/pull/956/files#r950306500)

File: apps/vscode/command_client/command_client.py
Comment (pokey) [inline apps/vscode/command_client/command_client.py:304]: yeah this name is not bad.  alternatively `run_ide_command...` or `run_editor_command...`.  But `rpc` seems reasonable

Comment (rntz) [reply apps/vscode/command_client/command_client.py:304]: I like `run_rpc_command`, it's appropriately general.

Comment (fidgetingbits) [reply apps/vscode/command_client/command_client.py:304]: I already started using this for something that's technically not an editor, so agree rpc is better

File: apps/vscode/command_client/command_client.py
Comment (fidgetingbits) [inline apps/vscode/command_client/command_client.py:17]: You could change this name to no longer include VSCODE

**Review by fidgetingbits — COMMENTED** [review]
This is nice, will be useful for a lot of different application plugins where keyboard shortcuts aren't enough. I think once the changes already requested are done, it will be good to go. I've already done some testing on a separate application and seems to work well.

File: apps/vscode/command_client/command_client.py
Comment (JohnEffo) [reply apps/vscode/command_client/command_client.py]: Double checked should be fixed.

File: apps/vscode/command_client/command_client.py
Comment (JohnEffo) [reply apps/vscode/command_client/command_client.py]: Double checked this is gone.

File: apps/vscode/command_client/command_client_tag.py
Comment (JohnEffo) [reply apps/vscode/command_client/command_client_tag.py]: Double checked it's gone.

File: apps/vscode/command_client/visual_studio.py
Comment (JohnEffo) [reply apps/vscode/command_client/visual_studio.py]: Gone.

File: apps/vscode/command_client/vscode.py
Comment (JohnEffo) [reply apps/vscode/command_client/vscode.py]: Gone

File: apps/vscode/command_client/command_client.py
Comment (JohnEffo) [reply apps/vscode/command_client/command_client.py:153]: Double checked this done, now :)

File: apps/vscode/command_client/command_client.py
Comment (JohnEfford) [reply apps/vscode/command_client/command_client.py:17]: Yes missed this sorry.

**Review by rntz — APPROVED** [review]

**Review by fidgetingbits — APPROVED** [review]

File: apps/vscode/command_client/command_client.py
Comment (pokey) [inline apps/vscode/command_client/command_client.py:16]: We should probably standardise on terminology here. I've been referring to VSCode as the server, not the client. Maybe "target application"? Cc/ @rntz @phillco

File: apps/vscode/command_client/command_client.py
Comment (pokey) [inline apps/vscode/command_client/command_client.py:295]:
```suggestion
        """Execute command via RPC."""
```

File: apps/vscode/command_client/command_client.py
Comment (pokey) [inline apps/vscode/command_client/command_client.py:344]:
```suggestion
        """Return the directory of the command server"""
```

File: apps/vscode/command_client/vscode.py
Comment (pokey) [inline apps/vscode/command_client/vscode.py:37]: This comment really applies to the last three actions. The first two will continue to be necessary so that we can still function without extension if they don't have it

File: apps/vscode/command_client/vscode.py
Comment (pokey) [inline apps/vscode/command_client/vscode.py:39]: What happened to the pre-phrase signal code? Unless I've missed something, it appears to have been dropped

**Review by pokey — CHANGES_REQUESTED** [review]
Looking pretty good; left a few more comments

File: apps/vscode/command_client/command_client.py
Comment (rntz) [reply apps/vscode/command_client/command_client.py:16]: Just "the application" seems fine here.

Comment (JohnEfford) [reply apps/vscode/command_client/command_client.py:16]: Removed client

File: apps/vscode/command_client/vscode.py
Comment (JohnEfford) [reply apps/vscode/command_client/vscode.py:37]: Moved to comment to correct position

File: apps/vscode/command_client/vscode.py
Comment (pokey) [reply apps/vscode/command_client/vscode.py:39]: Ah nvm I see it now. A bit confusing how it's declared in multiple places

What's the point of allowing applications to turn it off? Seems simpler to just leave it on and let all applications share the same impl

Comment (JohnEfford) [reply apps/vscode/command_client/vscode.py:39]: See command client 368 its in ctx which uses the match belwo to check we are in the correct context. I put a comment in command_client_tag to explain when to overwrite

```Python
ctx.matches = r"""
tag: command_client
"""
```

Comment (JohnEffo) [reply apps/vscode/command_client/vscode.py:37]: I'll remove it from the actions so that it cannot be turned of.

Comment (JohnEffo) [reply apps/vscode/command_client/vscode.py:37]: Done

File: apps/vscode/command_client/command_client.py
Comment (JohnEffo) [reply apps/vscode/command_client/command_client.py:295]: Done

File: apps/vscode/command_client/command_client.py
Comment (JohnEffo) [reply apps/vscode/command_client/command_client.py]: Done

File: apps/vscode/command_client/vscode.py
Comment (rntz) [reply apps/vscode/command_client/vscode.py:37]: Uh, this seems to have been changed into a non-working state. There's no module which declares `emit_pre_phrase_signal`, only a context which implements it, which should be producing an error in the talon log. We probably want the default implementation to be `return False`. There's no sense touching a file if we're not in a command client. I'll go change this now.

Comment (rntz) [reply apps/vscode/command_client/vscode.py:37]: Ok, this should be working now. @pokey is this okay? Did you want us to _always_ touch the pre-phrase file, even if we don't have a command-client app focused?

Comment (rntz) [reply apps/vscode/command_client/vscode.py:37]: interestingly, it doesn't seem to be causing an error in the talon log. not sure why.

Comment (rntz) [reply apps/vscode/command_client/vscode.py:37]: ok. it does error if I actually try to use the action, though. I'll try to minimize that and report it as a talon bug.

File: apps/vscode/command_client/command_client.py
Comment (rntz) [inline apps/vscode/command_client/command_client.py:34]: This is wrong! It needs to be `tag: user.command_client`. Ugh, talon should warn about this.

**Review by rntz — DISMISSED** [review]

File: apps/vscode/command_client/vscode.py
Comment (rntz) [reply apps/vscode/command_client/vscode.py:37]: ok, I've reported the underlying bug to talon, https://github.com/talonvoice/talon/issues/525.

File: apps/vscode/command_client/command_client.py
Comment (rntz) [reply apps/vscode/command_client/command_client.py]: fixed it.

**Comment by rntz** [comment]
ok, @pokey, I think this is ready for prime time. I've merged it with my local branch and it works for me. If you're ok with it I think we should merge.

File: apps/vscode/command_client/visual_studio.py
Comment (pokey) [inline apps/vscode/command_client/visual_studio.py:15]: Why the mixed format here?  I would argue for consistency with what we do for vscode:
```suggestion
        return "visual-studio-command-server"
```

**Review by pokey — COMMENTED** [review]
Ok I did some cleanup, and tested that things are still working for me locally.  I left one comment; otherwise good to go!

File: apps/vscode/command_client/visual_studio.py
Comment (JohnEfford) [reply apps/vscode/command_client/visual_studio.py:15]: Can you commit that for me, the location change seems fine to me.

**Comment by rntz** [comment]
@pokey I think this is just waiting on you to clear your "changes requested" review, until you do that I can't merge it.

**Review by pokey — APPROVED** [review]
~I'm happy, but let's give @JohnEffo a chance to review the [change to the server directory](https://github.com/knausj85/knausj_talon/pull/956/commits/3d1e01613da066798866c5a4504d6d92a241c4ac), as that will likely break his server implementation until he's done the rename Visual Studio-side~ whoops missed https://github.com/knausj85/knausj_talon/pull/956#discussion_r963807713

I'd also like to give @lunixbochs a chance to have a look before we merge, as this PR establishes the RPC mechanism that Visual Studio, VSCode, Emacs, and JetBrains will be using until there is an official Talon RPC, so it would be good to make sure we're not doing anything that will make our lives harder when we migrate to official Talon RPC

**Comment by pokey** [comment]
One other thought: we should probably move this directory in a follow-up PR, as it is still under `apps/vscode`.  Maybe let's wait for https://github.com/knausj85/knausj_talon/pull/897 though?

**Pull request merged**
